File: apps/web/lib/markdown-sections.ts
/**
 * The backend stores one `content_markdown` blob per ArtifactVersion — it
 * has no concept of "sections". The document editor UI is built around
 * sections (left-panel navigation, one editable block per section), so
 * these two functions reconcile the two: split real content into sections
 * by top-level (`##`) heading for editing, and join them back into one
 * blob before saving. Purely a client-side view convenience — nothing is
 * lost or reordered by round-tripping through these.
 */

import type { ArtifactSection } from "@/lib/types";

let sectionIdCounter = 0;
function nextSectionId(): string {
  sectionIdCounter += 1;
  return `section-${sectionIdCounter}`;
}

export function splitMarkdownIntoSections(markdown: string): ArtifactSection[] {
  const lines = markdown.split("\n");
  const sections: ArtifactSection[] = [];
  let currentTitle = "Content";
  let currentLines: string[] = [];
  let currentIsSynthetic = true;

  function flush() {
    const content = currentLines.join("\n").trim();
    if (content.length > 0 || !currentIsSynthetic) {
      sections.push({ id: nextSectionId(), title: currentTitle, contentMarkdown: content, isSynthetic: currentIsSynthetic });
    }
  }

  for (const line of lines) {
    const headingMatch = /^##\s+(.*)$/.exec(line);
    if (headingMatch) {
      flush();
      currentTitle = headingMatch[1].trim();
      currentLines = [];
      currentIsSynthetic = false;
    } else {
      currentLines.push(line);
    }
  }
  flush();

  // No "##" headings at all (e.g. a title-only "#" document, or plain
  // text) — treat the whole thing as one section rather than losing content.
  if (sections.length === 0) {
    return [{ id: nextSectionId(), title: "Content", contentMarkdown: markdown.trim(), isSynthetic: true }];
  }
  return sections;
}

/** True only if `markdown` has at least one genuine `##` heading —
 * mirrors apps/api/app/services/markdown_sections.py's has_real_sections.
 * splitMarkdownIntoSections's fallback "Content" section (above) is a
 * synthesized label that never appears in the source text, not a real,
 * individually-improvable section — see agent-actions-panel.tsx, which
 * uses this to disable "Improve section" for a headingless document
 * (asking the IMPROVE agent to revise "the Content section" there is
 * really asking it to regenerate the whole document from one
 * instruction, which has previously wiped out real content). */
export function hasRealSections(markdown: string): boolean {
  return /^##\s+\S/m.test(markdown);
}

// Mirrors apps/api/app/services/ai_generation.py's CLARIFICATION_MARKER —
// an agent's output starts with this exact heading when it needs more
// information before it can draft the real artifact (see
// format_clarification_output). Kept as a literal here rather than
// fetched from the backend since it's a fixed protocol constant, not
// configuration.
const CLARIFICATION_MARKER = "# Clarification Needed";

/** True if `markdown` is an agent's clarification request rather than a
 * real drafted document — see components/documents/clarification-panel.tsx
 * for the UI that lets a user answer it directly. */
export function isClarificationRequest(markdown: string): boolean {
  return markdown.trimStart().startsWith(CLARIFICATION_MARKER);
}

export function joinSectionsIntoMarkdown(sections: ArtifactSection[]): string {
  // A synthetic "Content" section's title never existed in the source —
  // writing it back as a literal "## Content" heading would permanently
  // inject a heading the original document never had (this exact bug
  // once turned a story backlog's intro paragraph into a "## Content"
  // section on its very first save). Write its text as-is instead.
  return sections
    .map((s) => (s.isSynthetic ? s.contentMarkdown : `## ${s.title}\n\n${s.contentMarkdown}`))
    .filter((part) => part.trim().length > 0)
    .join("\n\n");
}

File: apps/api/app/services/markdown_sections.py
"""Shared Markdown `##`-heading section helpers — mirrors
apps/web/lib/markdown-sections.ts exactly, and must stay in sync with it: a
section title is what round-trips a reviewer's comment
(ReviewComment.section_title) back to a specific part of a regenerated
document, and what app/services/graph_engine.py's
validate_evidence_requirement looks for (e.g. a required "Test Evidence"
section) before an artifact can be approved.

Split out of app/services/revision_agent.py (which still uses these for
its section-scoped merge) so app/services/graph_engine.py can use them too
without a circular import — revision_agent.py itself imports
GraphEngineService.
"""

import re

_HEADING_RE = re.compile(r"^##\s+(.*)$")


def split_into_sections(markdown: str) -> list[dict[str, str | bool]]:
    """Splits `markdown` into one dict per top-level (`##`) heading:
    `{"title": ..., "content": ..., "is_synthetic": ...}`, in document
    order. A document with no `##` headings at all (or text before its
    first one) comes back with that leading text as a "Content" section
    rather than losing it — `is_synthetic=True` marks that this "Content"
    title never actually appeared in the source, so join_sections (below)
    never turns it into a real, literal heading."""
    lines = markdown.split("\n")
    sections: list[dict[str, str | bool]] = []
    current_title = "Content"
    current_lines: list[str] = []
    current_is_synthetic = True

    def flush() -> None:
        content = "\n".join(current_lines).strip()
        if content or not current_is_synthetic:
            sections.append({"title": current_title, "content": content, "is_synthetic": current_is_synthetic})

    for line in lines:
        match = _HEADING_RE.match(line)
        if match:
            flush()
            current_title = match.group(1).strip()
            current_lines = []
            current_is_synthetic = False
        else:
            current_lines.append(line)
    flush()

    if not sections:
        return [{"title": "Content", "content": markdown.strip(), "is_synthetic": True}]
    return sections


def join_sections(sections: list[dict[str, str | bool]]) -> str:
    """Inverse of split_into_sections. A section flaged `is_synthetic`
    (the leading "Content" placeholder — never a key present when the
    section came from anywhere but split_into_sections's own fallback,
    so `.get` defaults False for a hand-built section) is written back as
    plain text, not `## Content` — that heading never existed in the
    original document and must not get invented by a save round-trip."""
    parts = []
    for s in sections:
        if s.get("is_synthetic"):
            parts.append(str(s["content"]).rstrip())
        else:
            parts.append(f"## {s['title']}\n\n{s['content']}".rstrip())
    return "\n\n".join(p for p in parts if p)


def has_real_sections(markdown: str) -> bool:
    """True only if `markdown` has at least one genuine `##` heading.
    `split_into_sections` synthesizes a single "Content" section (a title
    that never actually appears in the source text) when there are none —
    a meaningful distinction for any caller that needs to scope an edit to
    one *real* section: "improve the Content section" of a headingless
    document is not a scoped edit at all, it's the entire document, which
    the section-scoped agents aren't built to safely regenerate from a
    single instruction (see app/services/section_improve_agent.py)."""
    # _HEADING_RE is anchored with ^/$ and matched per-line elsewhere in
    # this module (no re.MULTILINE) — mirror that here rather than
    # .search() the whole string, which would only ever check the first line.
    return any(_HEADING_RE.match(line) for line in markdown.split("\n"))


def find_section(markdown: str, title: str) -> dict[str, str] | None:
    """Case-insensitive lookup of one section by heading title — used by
    graph_engine.py's validate_evidence_requirement to check a required
    section (e.g. "Test Evidence") actually exists and isn't empty."""
    normalized = title.strip().lower()
    for section in split_into_sections(markdown):
        if section["title"].strip().lower() == normalized:
            return section
    return None
